prusti-launch/src/bin/cargo-prusti.rs

```diff
@@ -35,8 +35,22 @@ where
     let prusti_log_dir =
         std::env::var("PRUSTI_LOG_DIR").unwrap_or_else(|_| format!("{}/log", cargo_target));
 
+    // If the target crate depends on the `prusti-contracts` crate, then enable its `prusti` feature.
```

Comment:
Why should the `prusti` feature enabled in that case?

Reply:
When running `cargo-prusti` or `prusti-rustc` I think that feature should always be enabled, which should be done by the following line

https://github.com/viperproject/prusti-dev/blob/3fbee9b0be249c051fa9f0b9ca8ff3be7327a324/prusti-launch/src/bin/prusti-rustc.rs#L122

Reply:
So this is a tricky one :smiley: Indeed `prusti-rustc` always enables the `prusti` feature. However, `prusti-rustc` is called with certain linking arguments for external crates. Since `prusti-contracts-internal` is now an optional dependency (which in turn explicitly depends on the `prusti-specs` crate and implicitly on the `prusti` crate), the compiler will fail due to error `error[E0433]: failed to resolve: use of undeclared crate or module 'prusti'` if `prusti-rustc` sees the `prusti` feature, because cargo does not add a link flag to this crate when invoking `prusti-rustc`.

Thus, the code here is to tell cargo to properly link the external crates on which `prusti-rustc` depends when the `prusti` feature is enabled.

Reply:
> What about always enabling --features=prusti-contracts/prusti? Have you found issues with that?

I've encountered two issues with this:

1) Cargo build times are significantly longer, not just the building of the source code, also `cargo doc`, `cargo clean` etc.
2) When turning on the `prusti-contracts/prusti` feature, this creates a dependency on the Rust standard library. This breaks compilation of `#![no_std]` crates when building for targets that don't support the Rust standard library.

> I'm trying to find an alternative that doesn't try to parse the Cargo.toml, because of whitespaces issues like with `prusti-contracts<space><space>=`.

This one I could work-around in the code, but I realized there are more pitfalls with my approach :smile: It's possible to specify [target specific dependencies](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#platform-specific-dependencies) in `Cargo.toml`, for example `[target.'cfg(windows)'.dependencies]`. The current implemented check would evaluate to true even if just 1 target depends on `prusti-contracts` but not the others. If the non-existing feature flag is enabled for targets which don't depend on `prusti-contracts`, this will result in an error.

Another corner case which seems relevant for issue #2, is that the current approach will not enable the prusti contracts for dependencies that are not a direct dependency specified in `Cargo.toml` (i.e. dependencies of dependencies).

Is there a way to make `prusti-rustc` always include the correct linking arguments, even if the cargo feature is not explicitly turned on? If that is possible, that might be a better solution.

Reply:
> Cargo build times are significantly longer, not just the building of the source code, also cargo doc, cargo clean etc.

Do you know if it's just because there are more dependencies to compile, or because the build cache gets invalidated because of the new flag? ([`cargo build -Ztimings`](https://internals.rust-lang.org/t/exploring-crate-graph-build-times-with-cargo-build-ztimings/10975/22) might be useful to debug build times.)

> When turning on the `prusti-contracts/prusti` feature, this creates a dependency on the Rust standard library. This breaks compilation of `#![no_std]` crates when building for targets that don't support the Rust standard library.

Removing any dependency on the standard library from `prusti_contracts` looks difficult. We could ask the developers to add a target that supports the standard library just for the Prusti verification. Would that be too restrictive?

> Is there a way to make prusti-rustc always include the correct linking arguments, even if the cargo feature is not explicitly turned on? If that is possible, that might be a better solution.

That looks difficult because of the random-looking filenames.

Reply:
> Do you know if it's just because there are more dependencies to compile, or because the build cache gets invalidated due to the new flag?

Without the `prusti` feature for `prustic-contracts`, it compiles 7 dependencies in total for a crate where `prusti-contracts` is the only dependency. When the `prusti` feature is enabled, this number goes up to 65 dependencies. 

> Removing any dependency on the standard library from prusti_contracts looks difficult. We could ask the developers to add a target that supports the standard library just for the Prusti verification. Would that be too restrictive?

Another alternative I see here, is that we always enable `prusti` feature by default for `prusti-contracts`, but it can be disabled in the host crate's `Cargo.toml` like so:

```toml
[dependencies]
prusti-contracts = { path = "path/to/prusti-dev/prusti-contracts", default-features = false }
```

and then only explicitly enable it when needed, e.g.:

```shell
cargo-prusti --features prusti-contracts/prusti
```

This still gives the performance benefits when not building with prusti (e.g. when building a crate obtained from crates.io) and avoids pulling in the Rust standard library, but requires slightly more management on the end-user side in order to obtain these benefits.

This doesn't fully address cross-crate verification using Prusti (also see #2), but this can be worked-around by the host crate's `Cargo.toml`:

```toml
[depedencies]
prusti-contracts = { path = "path/to/prusti-dev/prusti-contracts", default-features = false }

[features]
default = []
prusti = ["prusti-contracts/prusti"]
```

and then another crate depending on this crate, can enable its `prusti` feature for cross-crate verification (this again needs a little extra management by the crate maintainer and would probably need to be documented somewhere).

I was hoping for a zero-configuration solution that always works(tm), although that seems non-trivial to achieve :) This also seems particularly tricky to implement in just the `cargo_verify/test_no_std` library crate, because the `prusti` feature should be off when running `cargo build`, but it should be on when running `cargo-prusti` and I'm not sure about what is the cleanest way to implement this. Ideally without changes to `cargotest.rs`, but I don't see another way currently.

> That looks difficult because of the random-looking filenames.

One way to do this might be to do a glob pattern search in the prusti target directory for `prusti-contracts*.rlib` and sort the artifacts by last modification time. And then only link the last touched library file from `prusti-rustc`, because this should be the most recently built library file. Likewise for other libraries that might need to be linked. Would this work for all intended usages of Prusti? Or couldn't we assume that these library files are present on the file systems in all thinkable use cases of Prusti (now or in the future)? Also there might be exotic systems out there that do not set the last modification time of a file (e.g. when the [`S_NOCMTIME`](https://serverfault.com/a/906970) kernel flag is set, see [here](https://www.kernel.org/doc/html/v4.14/filesystems/index.html#c.file_update_time)). Although, these are most likely optimized production systems and not build servers, so I'm not sure if this will be an issue in practice (quite a footgun still for someone that wants to verify their crate using Prusti on such a server :P ).

Reply:
I would avoid trying the `prusti-contracts*.rlib` glob pattern search or checking the modification times. So far we don't depend on those cargo internals.

> Without the `prusti` feature for `prustic-contracts`, it compiles 7 dependencies in total for a crate where `prusti-contracts` is the only dependency. When the `prusti` feature is enabled, this number goes up to 65 dependencies.

There is still something that I don't understand. Why having `enable_prusti_contracts` always `true` would cause more dependencies to be compiled? If Prusti is running and if a crate depend on `prusti_contracts`, I expect that those extra dependencies that gets compiled are necessary to run `prusti_contracts_internal`. I don't see how we can avoid them. If Prusti is not running or if the crate does not depend on `prusti_contracts`, then I expect that the dependencies of `prusti_contracts_internal` don't get compiled anyway.

Reply:
> If Prusti is running and if a crate depend on prusti_contracts, I expect that those extra dependencies that gets compiled are necessary to run prusti_contracts_internal. I don't see how we can avoid them.

Agreed :)

> If Prusti is not running or if the crate does not depend on prusti_contracts, then I expect that the dependencies of prusti_contracts_internal don't get compiled anyway.

This is indeed the effect I'm trying to achieve. If the `prusti` feature is enabled for the `prusti-contracts` dependency, then all 65 dependencies will be compiled regardless if the build is invoked by `cargo build` (the non-Prusti version) or `cargo-prusti`. It also affects the evaluation of the macros in `prusti-contracts`. Unfortunately, the vanilla `cargo` is not able to automatically detect that compiling the additional dependencies or evaluating the macros from `prusti-contracts-internal` are not needed, when the `prusti` feature is enabled by default in `prusti-contracts`. One work-around would be as suggested in my previous comment by specifying `default-features =  false` for the `prusti-contracts` dependency in the crate of a Prusti end-user. However, this puts an additional burden on the end-user to get this right. So from the perspective of a better end-user experience, I'm still searching for a "nicer" way (although, if we at some point indeed exhaust all options, it is a viable alternative).

I have one more idea I want to try today. It's possible to produce different kinds of libraries through setting the [`crate-type` ](https://doc.rust-lang.org/reference/linkage.html) in `Cargo.toml`. I'm hoping that one of these options will produce a linkable artifact which has a deterministic file name, in order to avoid extra trickery in the `cargo-prusti` internals. This artifact will then be linked through `prusti-rustc` and we can leave `cargo-prusti` untouched if I can get it to work. If this works, then we can leave the `prusti` feature disabled by default in `Cargo.toml` for `prusti-contracts`, but then it will still properly compile `prusti-contracts-internal` when passing through `prusti-rustc`. Hopefully I can show more later today :)

Reply:
> This is indeed the effect I'm trying to achieve.

Thanks! Sorry for asking trivial questions; I'm trying to evaluate all options without actually trying them because I don't have much spare time :)

> If the prusti feature is enabled for the prusti-contracts dependency, then all 65 dependencies will be compiled regardless if the build is invoked by cargo build (the non-Prusti version) or cargo-prusti

From the cargo documentation "by default, all features are disabled unless explicitly enabled." So, how is the prusti feature enabled by default in `prusti-contracts/Cargo.toml`? I don't see any `default = [...]` in it.

Reply:
No worries! Happy to explain in further depth, because this also strengthens my own understanding of what is going on :)

> So, how is the prusti feature enabled by default in prusti-contracts/Cargo.toml?

Before this PR, `prusti-contracts-internal` was not an optional dependency, hence it was always included in the linked libraries, including its 65 dependencies:

https://github.com/viperproject/prusti-dev/blob/75eec658bfc8fe716d8f465d30bf4d6eb5b12012/prusti-contracts/Cargo.toml#L7-L9

In the current state of this PR, there is indeed no `default = ["prusti"]` in `prusti-contracts/Cargo.toml`. Instead, the enabling of the `prusti` feature is injected through `cargo-prusti`:

https://github.com/viperproject/prusti-dev/blob/dcb95e391ca0ea67513de9c68c5e58dcd0af4e54/prusti-launch/src/bin/cargo-prusti.rs#L38-L52

> What about always enabling `--features=prusti-contracts/prusti`? Have you found issues with that?

I just realized I may have misinterpreted this earlier question and did not give a proper full explanation :) One piece of information that was missing in my explanation is that this will break the build if the crate does not depend on `prusti-contracts` with `` error: Package `my_crate` does not have a dependency named `prusti-contracts` ``. This is quite often the case, for example when running Prusti on crates from crates.io as described in one of the Prusti papers. So the above linked code will do a check if the crate depends on `prusti-contracts`, and if so, then inject the `--features=prusti-contracts/prusti` snippet in the call to `cargo`.

However, we also established that this approach in `cargo-prusti` is not ideal for these reasons:

1. whitespaces issues like with `prusti-contracts<space><space>=`;
2. the `prusti-contracts` dependency may be conditional on the target platform;
3. it won't work when (in the future) we run Prusti cross-crate, where we also want to evaluate the specifications of dependencies (of dependencies, and so on).

The solution that I'm working towards, is that the `prusti` feature of `prusti-contracts` is never enabled in any `Cargo.toml` nor in `cargo-prusti`. Instead, the `--cfg=feature=\"prusti\"` flag is injected by `prusti-rustc` (this is already there today, like you linked earlier). This will correctly trigger the macro expansion from `prusti-contracts-internal`. But, because `prusti-contracts-internal` is an optional dependency, cargo does not link the `prusti-contracts-internal` crate into the `prusti-rustc` call. This creates another compilation error, because it can't find `prusti-contracts-internal`. To address this last problem, I will attempt to update `prusti-rustc` to always explicitly link `prusti-contracts` and `prusti-contracts-internal`, even if the crate has no dependency on `prusti-contracts`.

Reply:
> Instead, the enabling of the `prusti` feature is injected through `cargo-prusti`:

Thanks, I thought you meant that it's enabled by cargo, but it's cargo-prusti.

> will break the build if the crate does not depend on `prusti-contracts`

Aha! Yes, I was missing that. Now I fully understand the messages.

> To address this last problem, I will attempt to update `prusti-rustc` to always explicitly link `prusti-contracts` and `prusti-contracts-internal`, even if the crate has no dependency on `prusti-contracts`

Even if you manage to generate the correct `rustc` arguments to link `prusti-contracts` and `prusti-contracts-internal`, what guarantees that they have been compiled?

Reply:
> Even if you manage to generate the correct rustc arguments to link prusti-contracts and prusti-contracts-internal, what guarantees that they have been compiled?

It is currently an assumption that someone will have run the command `x.py build` first :) I believe it is currently already the case on the `master` branch that we have to run `x.py build` before being able to run `x.py test`. For example, this fails:

```shell
$ x.py clean
$ x.py setup
$ x.py test
--- SNIP ---
     Running tests/cargotest.rs (target/debug/deps/cargotest-c8ee7120bcf841fd)

running 6 tests
test simple_assert_false ... FAILED
test simple_assert_true ... FAILED
test test_failing_crate ... FAILED
test test_prusti_toml ... FAILED
test test_prusti_toml_fail ... FAILED
test test_symlinks ... ok
```

All the tests failed on the same error `thread 'main' panicked at 'failed to execute prusti-driver ("/path/to/prusti-dev/target/debug/prusti-driver")', prusti-launch/src/bin/prusti-rustc.rs:135:29`. If I run `x.py build` before `x.py test`, this error does not happen.

> Would this work for all intended usages of Prusti? Or couldn't we assume that these library files are present on the file systems in all thinkable use cases of Prusti (now or in the future)?

Although, come to think of it, this is not a hard enough guarantee when Prusti is shipped in binary form (e.g. for the Prusti Assistant or through [Github releases](https://github.com/viperproject/prusti-dev/releases)). Would it be possible to include the linked libraries as separate files with the binary distributions of Prusti? Are there any other corner cases where this might be a problem?

Reply:
Was just digging little deeper and found out that there is already a `libprusti_contracts.rlib` shipped with Prusti and used in:

https://github.com/viperproject/prusti-dev/blob/697ee88f78fc2b8de9ded763b362e9722836b0e8/prusti-launch/src/bin/prusti-rustc.rs#L111-L119

Right now it's just missing the `.rlib` for `prusti-contracts-internal` (a `proc-macro` crate). I'm going to investigate further if we can bring that in the same way as `prusti-rustc` already does for `prusti-contracts`.

Reply:
> I believe it is currently already the case on the `master` branch that we have to run `x.py build` before being able to run `x.py test`

To test Prusti yes, but I was thinking at use cases like VSCode.

> Would it be possible to include the linked libraries as separate files with the binary distributions of Prusti? Are there any other corner cases where this might be a problem?

Yes, good point! Including the libraries in the release is certainly possible. Actually, we already release some of them (I forgot about it), but I think they are never used by Prusti. That way, the library needs to be chosen when we prepare the release and that makes the glob pattern approach more robust (e.g. we don't have to worry about weird file systems or cargo version; we control them).

The only question is what happens if the source code of the `prusti_contracts`/`prusti_contracts_impl` dependency used by the crate to be verified doesn't match the library in the release. Like, what if we add or remove a derive macro? How bad is the cargo error message (if any)? Ideally, Prusti should give a readable error message, but if the cargo error is not too bad we could postpone the readable error to a future PR.

Reply:
@vakaras I saw in the git history that you've worked on this section in the past, might you have any tips how to best explicitly link the `prusti-contracts-internal` proc-macro crate from within `prusti-rustc`? :) In this PR, I've made `prusti-contracts-internal` an optional dependency of `prusti-contracts`. `prusti-contracts-internal` is still linked correctly though:

```shell
$ rustc -Z ls target/debug/libprusti_contracts.rlib
Crate info:
name prusti_contracts-410d1d3be9361555
hash 910f03a0d2f44dca stable_crate_id StableCrateId(10652995453112551957)
proc_macro false
=External Dependencies=
1 core-0e3656b1fda5fd7b hash 90110cb6a5230b89 host_hash None kind Explicit
2 compiler_builtins-eecefd843a0dbc02 hash 7951e33a8234605c host_hash None kind Explicit
3 rustc_std_workspace_core-2a6a2797f7a73818 hash df1b6930d614cd5e host_hash None kind Explicit
4 prusti_contracts_internal-4e86e2685f664ef2 hash 214fce9f03f0ad0d host_hash None kind MacrosOnly
```

And when I run `prusti-rustc` I get an error: `` use of undeclared crate or module `prusti_contracts_internal` ``.

Things I've tried so far:

* Add a `crate-type = ["rlib"]` in `prusti-contracts-internal/Cargo.toml`, but then cargo complains about `` cannot mix `proc-macro` crate type with others ``.
* I considered adding an extra intermediary crate between `prusti-contracts` and `prusti-contracts-internal` to work-around the above error message, but I'm worried that this intermediary crate will then again depend on `prusti-contracts-internal` and get the same error. I still have to try this though.

Reply:
> Including the libraries in the release is certainly possible. Actually, we already release some of them (I forgot about it), but I think they are never used by Prusti.

Looks like one of those is used already. `prusti-rustc` uses `libprusti_contracts.rlib` currently.

> The only question is what happens if the source code of the `prusti_contracts`/`prusti_contracts_impl` dependency used by the crate to be verified doesn't match the library in the release. Like, what if we add or remove a derive macro? How bad is the cargo error message (if any)? Ideally, Prusti should give a readable error message, but if the cargo error is not too bad we could postpone the readable error to a future PR.

When removing a macro, we could replace the token stream with a `compile_error!("macro X has been removed/renamed in version Y of Prusti");` [statement](https://doc.rust-lang.org/std/macro.compile_error.html). I don't foresee problems with adding a new macro. If the end-user is on an older version of `prusti-contracts-impl`, that new macro won't exist yet and they won't be able to call it yet (barring any name collisions with other imports when new Prusti macros are introduced, but the Rust compiler will give a decent error message in that case).

In the case the versions mismatch, I don't expect any linking errors. Because the distributed `.rlib` file is always on the same version as the Prusti binaries.

Reply:
When doing cleans build of Prusti:

```
x.py clean
x.py build
x.py build --release
```

I see the following libraries for `prusti-contracts-internal`:

```
target/debug/deps/libprusti_contracts_internal-4e86e2685f664ef2.so
target/debug/libprusti_contracts_internal.d
target/debug/libprusti_contracts_internal.so
target/release/deps/libprusti_contracts_internal-a2ff594f86b4a9d7.so
target/release/deps/libprusti_contracts_internal-0f18580e1dd0de70.so
target/release/libprusti_contracts_internal.d
target/release/libprusti_contracts_internal.so
```

In the debug output, there is only 1 `.so` file in the `deps/` folder, but for the release build, it has 2 `.so` files in its `deps/` folder. Is there a specific reason for that? I would expect only 1 `.so` file in both builds, but I'm venturing into new territory here so I'm not sure if that's intentional or not :)

Reply:
Quick update: I managed to find a relatively simple fix for the linking of `prusti-contracts-internal`. I still have to fix a test case on Windows for `cargo_verify/test_no_std` and then I'll do a write-up of what I found out :) This might take me a couple of more commits before the Windows test is fixed, as I don't have access to a Windows machine atm.